File: collapsingtoolbarlayout-subtitle/tests/src/com/hendraanggrian/collapsingtoolbarlayout/subtitle/test/SimpleViewAction.java
package com.hendraanggrian.collapsingtoolbarlayout.subtitle.test;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.test.espresso.UiController;
import android.support.test.espresso.ViewAction;
import android.view.View;

import org.hamcrest.Matcher;

import static android.support.test.espresso.matcher.ViewMatchers.isAssignableFrom;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public abstract class SimpleViewAction<T extends View> implements ViewAction, OnPerform<T> {

    @NonNull private final Class<T> viewCls;
    @Nullable private final String desc;

    SimpleViewAction(@NonNull Class<T> viewCls) {
        this.viewCls = viewCls;
        this.desc = null;
    }

    SimpleViewAction(@NonNull Class<T> viewCls, @NonNull String desc) {
        this.viewCls = viewCls;
        this.desc = desc;
    }

    SimpleViewAction(@NonNull Class<T> viewCls, @NonNull String format, @NonNull Object... formatArgs) {
        this.viewCls = viewCls;
        this.desc = String.format(format, formatArgs);
    }

    @Override
    public Matcher<View> getConstraints() {
        return isAssignableFrom(viewCls);
    }

    @Override
    public String getDescription() {
        return desc;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void perform(UiController uiController, View view) {
        onPerform((T) view);
    }
}

File: collapsingtoolbarlayout-subtitle/tests/src/com/hendraanggrian/collapsingtoolbarlayout/subtitle/test/SimpleViewAction.kt
package com.hendraanggrian.collapsingtoolbarlayout.subtitle.test

import android.support.test.espresso.UiController
import android.support.test.espresso.ViewAction
import android.support.test.espresso.matcher.ViewMatchers.isAssignableFrom
import android.view.View

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
abstract class SimpleViewAction<T : View> @JvmOverloads constructor(
        val viewCls: Class<T>,
        val desc: String? = null,
        vararg val formatArgs: Any = emptyArray()
) : ViewAction {

    abstract fun onPerform(t: T)

    override fun getConstraints() = isAssignableFrom(viewCls)

    override fun getDescription() = desc

    @Suppress("UNCHECKED_CAST")
    override fun perform(uiController: UiController, view: View) = onPerform(view as T)
}